### PR TITLE
diff calculator

### LINKED ISSUE #12: diff results

### PR DESCRIPTION
**Pull request by DavidBadura**
you can diff the analyse result between two versions. 

```bash
bin/simpspector diff [from(path|result)] [to(path|result)]
```

to save results i add an option for command `analyse` named `-o|--output`.

```bash
bin/simpspector analyse ./ -f json -o result.json
```

as example:

```bash
# checkout master branch
git checkout master

# analyser project and save it in result.json
bin/simpspector analyse ./ -f json -o result.json

# checkout new feature
git checkout new-feature

# diff result between result.json and workspace
bin/simpspector diff result.json ./
```

copy from https://github.com/simpspector/simpspector/pull/98
fixed #12

### CODE REVIEW

**Comment by tolry** [comment]
Is this ready for review and merging or still WIP?

**Comment by DavidBadura** [comment]
This is still WP, sorry^^ but a review is welcome. :wink:

**Comment by DavidBadura** [comment]
btw i wanted add a simple scoring system but it escalated quickly :D

File: src/Result.php [hunk 1]
```diff
@@ -18,13 +20,20 @@ class Result
...
 
     /**
+     * @var float
+     */
+    private $score;
+
+    /**
      * @param Issue[] $issues
      * @param Metric[] $metrics
+     * @param float|int $score
```
Comment (tolry) [inline src/Result.php]: Do we want to allow float and int, not restrict to int [0..100]. Another thing: I thought the metrics already could include data for score calculation, right? We had some big flexible system discussed on last month's coding dojo, but I honestly don't remember everything in detail anymore. Maybe talk about this later?

File: src/Executor/Executor.php [hunk 2]
```diff
@@ -58,6 +70,11 @@ public function run($path, array $config, AbstractLogger $logger = null)
         $result = $this->executeGadgets($path, $config, $logger);
 
         $event = new ExecutorResultEvent($path, $config, $result, $logger);
+        $this->eventDispatcher->dispatch(Events::BEFORE_SCORE, $event);
+
+        $result->setScore($this->scoreCalculator->calculate($result));
```
Comment (tolry) [inline src/Executor/Executor.php]: OK, forget about my other comment regarding the score attribute :smile: The 'problem' for me with the result object as well as the issues is, that there is some kind of hidden contract in them. Some data should be entered by the different gadgets, some data will be automatically set by the executor or EventListeners/Subscribers and its not obvious 'whats what'. Gadgets do not really know what data they can set and which is replaced later.

**Comment by tolry** [comment]
This is quite a big PR. For the future I would strongly suggest keeping these smaller. But this is important stuff that we need to get on the road, so ...

I would really like to use aliases on the use statements for the two distinct Result objects, that would have saved me a lot of confusion while reviewing this code

```php
use SimpSpector\Analyser\Result as AnalyserResult;
use SimpSpector\Gadget\Result as GadgetResult;
```

Still I would like to have a quick walk through today before evaluating the changes any further.

**Comment by DavidBadura** [comment]
will be moved to inspector

**Comment by DavidBadura** [comment]
only score calculator will be moved to inspector.

**Comment by DavidBadura** [comment]
rebase & remove scoring

**Comment by DavidBadura** [comment]
update description

**Comment by DavidBadura** [comment]
ready to merge

**Comment by tolry** [comment]
:+1: 

no time to review - we can refactor everything on the hackday anyway :smile:

**Pull request merged**